source/common/http/conn_manager_impl.cc

```diff
@@ -1529,7 +1524,63 @@ void ConnectionManagerImpl::ActiveStream::sendLocalReply(
     state_.saw_connection_close_ = true;
   }
 
-  stream_info_.setResponseCodeDetails(details);
+  if (response_headers_.get() == nullptr) {
+    // If the response has not started at all, send the response through the filter chain.
+    sendLocalReplyViaFilterChain(is_grpc_request, code, body, modify_headers, is_head_request,
+                                 grpc_status, details);
+  } else if (!response_headers_encoded_) {
+    ENVOY_STREAM_LOG(debug, "Sending local reply with details {} directly to the encoder", *this,
+                     details);
+    // In this case, at least the header and possibly the body has started
+    // processing through the filter chain, but no non-informational headers
+    // have been sent downstream. To ensure that filters don't get their
+    // state machine screwed up, bypass the filter chain and send the local
+    // reply directly to the codec.
+    //
+    // Make sure we won't end up with nested watermark calls from the body buffer.
+    state_.encoder_filters_streaming_ = true;
+    Http::Utility::sendLocalReply(
+        state_.destroyed_,
+        Utility::EncodeFunctions{
+            [&](ResponseHeaderMap& response_headers, Code& code, std::string& body,
+                absl::string_view& content_type) -> void {
+              connection_manager_.config_.localReply().rewrite(
+                  request_headers_.get(), response_headers, stream_info_, code, body, content_type);
+            },
+            [&](ResponseHeaderMapPtr&& response_headers, bool end_stream) -> void {
+              if (modify_headers != nullptr) {
+                modify_headers(*response_headers);
+              }
+              response_headers_ = std::move(response_headers);
+              encodeHeadersInternal(*response_headers_, end_stream);
+            },
+            [&](Buffer::Instance& data, bool end_stream) -> void {
+              encodeDataInternal(data, end_stream);
+            }},
+        Utility::LocalReplyData{Grpc::Common::hasGrpcContentType(*request_headers_), code, body,
+                                grpc_status, state_.is_head_request_});
```

Comment:
This is mostly copied from below and is somewhat nuanced and took me a couple of reads to see the difference. Is there any way to share this with just a bool that switches between internal vs. not? Up to you.